### PR TITLE
Feat: lookback windows closes #728

### LINKED ISSUE #728: Add Lookback Window
This solves a late arriving data use case. The lookback window is defined as a model property for incremental models and define the number of cadences to subtract from the start timestamp when doing a cadence loads (ignored on backfills). Ex: A daily model with a lookback window of 2 would subtract 2 days from the start date.

### PR DESCRIPTION
**Pull request by tobymao**

### CODE REVIEW

File: sqlmesh/core/model/meta.py [hunk 1]
```diff
@@ -216,20 +217,18 @@ def _date_validator(cls, v: t.Any) -> t.Optional[TimeLike]:
...
         else:
-            batch_size = int(v.name)
-        if batch_size is not None and batch_size <= 0:
-            raise ConfigError(
-                f"Invalid batch size {batch_size}. The value should be greater than 0"
-            )
-        return batch_size
+            num = int(v.name)
+        if num is not None and num <= 0:
+            raise ConfigError(f"Invalid value {num}. The value should be greater than 0")
```
[the remark this answers is not in the document]
Comment (tobymao) [reply sqlmesh/core/model/meta.py:230]: thanks

File: sqlmesh/core/model/meta.py [hunk 2]
```diff
@@ -238,8 +238,9 @@ def _kind_validator(cls, values: t.Dict[str, t.Any]) -> t.Dict[str, t.Any]:
             if not kind.is_materialized:
                 if values.get("partitioned_by_"):
                     raise ValueError(f"partitioned_by field cannot be set for {kind} models")
-            if values.get("batch_size") and not kind.supports_batch_size:
-                raise ValueError(f"batch_size field cannot be set for {kind} models")
+            for prop in ("batch_size", "lookback"):
+                if values.get(prop) and not kind.is_incremental:
+                    raise ValueError(f"{prop} field cannot be set for {kind} models")
```
Comment (crericha) [inline sqlmesh/core/model/meta.py:243]: Maybe provide more context. `{prop} field only works with incremental models, cannot be set for {kind} models`

File: sqlmesh/core/snapshot/definition.py [hunk 3]
```diff
@@ -504,26 +505,40 @@ def missing_intervals(
...
         )
-        dates = tuple(croniter_range(start_dt, end_dt, self.model.normalized_cron()))
-        size = len(dates) - 1
 
-        for i in range(size):
+        croniter = self.model.croniter(start_dt)
+        dates = [start_dt, to_datetime(croniter.get_next())]
+
+        while dates[-1] < end_dt:
+            dates.append(to_datetime(croniter.get_next()))
```
Comment (eakmanrq) [inline sqlmesh/core/snapshot/definition.py:519]: Can you add some comments to this code to better explain what is going on?

File: docs/concepts/models/overview.md [hunk 4]
```diff
@@ -94,6 +94,9 @@ An optional arbitrary string sequence used to create new model versions without
 ### batch_size
 - Batch size is used to optimize backfilling incremental data. It determines the maximum number of intervals to run in a single job. For example, if a model specifies a cron of `@hourly` and a batch_size of `12`, when backfilling 3 days of data, the scheduler will spawn 6 jobs. (3 days * 24 hours/day = 72 hour intervals to fill. 72 intervals / 12 intervals per job = 6 jobs.)
 
+### lookback
+- Lookback is used for incremental models to capture late arriving data. This must be a positive integer and refers to the number of units that late arriving data is expected.
```
Comment (izeigerman) [inline docs/concepts/models/overview.md:98]: Let's add link to the `INCREMENTAL_BY_TIME_RANGE` kind here when referring to `incremental`.

File: sqlmesh/core/model/meta.py [hunk 5]
```diff
@@ -50,7 +51,9 @@ class ModelMeta(PydanticModel):
     description: t.Optional[str]
     stamp: t.Optional[str]
     start: t.Optional[TimeLike]
+    retention: t.Optional[int]  # not implemented yet
```
Comment (izeigerman) [inline sqlmesh/core/model/meta.py:54]: Why add this now?

File: tests/core/test_snapshot.py [hunk 6]
```diff
@@ -210,6 +210,80 @@ def test_missing_intervals(snapshot: Snapshot):
...
+    ]
+    assert snapshot.missing_intervals("2023-01-28", "2023-01-28", "2023-01-30 05:00:00") == []
+    assert snapshot.missing_intervals("2023-01-29", "2023-01-29", "2023-01-30 05:00:00") == []
+    assert snapshot.missing_intervals("2023-01-27", "2023-01-29", "2023-01-30 05:00:00") == [
+        (to_timestamp("2023-01-27"), to_timestamp("2023-01-28")),
+    ]
+
+    assert snapshot.missing_intervals("2023-01-30", "2023-01-30", "2023-01-30") == [
+        (to_timestamp("2023-01-30"), to_timestamp("2023-01-31")),
+    ]
```
Comment (crericha) [inline tests/core/test_snapshot.py:284]: Add test case for end > latest

**Pull request merged**